### DISCUSSION TITLE
Ability to choose when to optimize images

### DISCUSSION THREAD

**Discussion by y-nk** [withastro/roadmap#760] [Proposal]
### Body

# Summary

TLDR: this RFC propose upgrading the `@astro/assets` feature to be on par with the different modes which Astro allows.

# Background & Motivation

I'm doing a website with `output: 'hybrid'` in my config because of many reasons (tons of static pages generated, but some not). In that configuration, even for images being part of the "runtime" pages, all images used with `@astro/assets` are preprocessed at build.

I understand why this happens as I've studied a bit the codebase.

Imho this is something that needs to evolve since the @astro/assets rendering strategy is not on par with the different modes that astro itself provides. We can summarize like this:

| output   | page rendering strategy | image rendering strategy | on par |
|----------|-------------------------|--------------------------|--------|
| `static` | at build                | at build                 | YES.   |
| `hybrid` | at build or at runtime  | at build                 | NO.    |
| `server` | at runtime              | at runtime               | YES.   |

# Goals

**For astro:assets to be on par with the rendering modes of the routes**

# Implementation

As discussed with @Princesseuh over Discord, there are a couple of things which need to be discussed, as I think the API fashion of Astro cannot be transposed to assets easily.

For example, in hybrid mode if you were to opt-out for static rendering, the way is to add `export const prerender = false` in the frontmatter of a `.astro` file. Of course, since assets don't have frontmatter, or even aren't astro files... then we can't do that.

Also, there could be the case where an image is willingly rendered statically at first (giving a specific couple of width/height, like thumbnails of a gallery) but larger versions could be rendered dynamically afterwards.

For these use-cases it is clear that the translation of Astro's `export const prerender = false` has to be a parameter of the `getImage()` function (and therefore a prop of `<Image />`

About the parameter itself, as discussed with @Princesseuh it should be "something along the line of `prerender: boolean`".

In my PR I implemented it as `transform: "eager" | "lazy"` to mimick `HTMLImageElement['loading']` parameter which also is a `"eager" | "lazy"` one.

There's no further complication as far as I know, but of course I could be wrong (that is why PR review exists)

# Proposal

https://github.com/withastro/astro/pull/8904

**Comment by y-nk** [comment]
@ematipico @Princesseuh applied RFC as requested in the PR

**Comment by y-nk** [comment]
On a side note, I've looked into forking or creating my own of this "in the meantime" (because our prod needs to move forward, but i understand also Astro's PoV) but I think it's not possible since there's a "leak" of `@astro/assets` in the core (with `globalThis.astroAsset.addStaticImage`)

From reading the code I understand why it was done, and I don't have sufficient knowledge to think of a better way, but I can see that this way of implementing (clever and efficient) sadly does not allow replacement by a 3rd party module.

_For now_, in order to let Astro take the time it needs and not block my work I've opted for a `patch-package` approach, hoping this will land sometime in the future.

**Comment by Princesseuh** [comment]
I like this and want it, honestly the only thing I don't know is how to present it in a way where it's not confusing for users of static and server.

Hybrid is our least used mode, so it's likely that an option that only works for it, while naming it something too general will lead to people being confused when it doesn't work. And sure, we can add warnings and errors and everything, but I'd rather try to think if there's a clever naming or something else to make it more clear

**Comment by ematipico** [comment]
Thank you @y-nk for putting time into creating this RFC.

I think the main goal of this proposal is 
> For `astro:assets` to be on par with the **rendering** modes of the routes

So, the expectations are the same. Could you please update them?

The table you created is fantastic and explains very well the shortcomings of the current behaviour of `astro:assets`.

> In my PR I implemented it as transform: "eager" | "lazy" to mimick HTMLImageElement['loading'] parameter which also is a "eager" | "lazy" one.

What do "eager" and "lazy" do when they are used with a `<img />` tag? Please ELI5 these options for me,  assume that I am a newbie. And how do these options change the behaviour of the current assets against the output modes? The RFC doesn't explain that.

**Reply by y-nk** [reply]
@ematipico sorry for the long delay (i'm overwelhmed with work atm)

> So, the expectations are the same. Could you please update them?

done!

> What do "eager" and "lazy" do when they are used with a <img /> tag?

Here's the best ELI5 on the topic: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#loading

The `<img />` tag has an existing `loading: 'eager' | 'lazy'` property which defines when the image will be loaded (either eagerly at page load, or lazily when the viewport reaches it).

I proposed (probably wrongly - i'm bad at naming) to have a similar interface with `transform` where `eager` would mean "at build time" and `lazy` "at runtime" so that it'd be similar in fashion... but maybe that's confusing?

**Reply by ematipico** [reply]
What do you think if we named the new option closer to what **Astro** does instead? e.g. `prerender: true | false` because that's the RFC is trying to solve. I fear that using something closer to HTML but with a different semantic could cause confusion in the short run.

**Reply by y-nk** [reply]
@ematipico i've no opinion, it's totally fine for me. i understand your point that it could confuse people, i can see that too.

On the "pros" side of your naming, i can see that the naming will be matching the page `export const prerender: boolean` api, so it looks nice from this PoV.

On the "soft cons" side, it means we'll have to write `<Image prerender={false} />` to "enable" the feature. I personally think/prefer that boolean props should be defaulted to false and enableable by writing their name only (like most of html props, for example `<input type="checkbox" checked />`)

...but i can live with that :)

if that's ok for you and @Princesseuh i can reflect the changes on my PR with `prerender: boolean`.

**Reply by y-nk** [reply]
bump! @ematipico 🙏

**Reply by ematipico** [reply]
Ooooh, I see. I think then the first proposal can work :)
